### ISSUE TITLE
Eorror when use load_checkpoint_and_dispatch: module 'torch.cuda' has no attribute 'mem_get_info'

### ISSUE THREAD

**Issue by BruceStayHungry** [huggingface/accelerate#880]
### System Info

```Shell
- `Accelerate` version: 0.13.2
- Platform: Linux-5.15.0-52-generic-x86_64-with-glibc2.17
- Python version: 3.8.13
- Numpy version: 1.23.3
- PyTorch version (GPU?): 1.9.0+cu102 (True)
```

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [ ] My own task or dataset (give details below)

### Reproduction

When trying to load gpt-neox-20B, I had the following error:
```
File ~/anaconda3/envs/intentQ/lib/python3.8/site-packages/accelerate/big_modeling.py:352, in load_checkpoint_and_dispatch(model, checkpoint, device_map, max_memory, no_split_module_classes, offload_folder, offload_buffers, dtype, offload_state_dict, preload_module_classes)
    344     max_memory = get_balanced_memory(
    345         model,
    346         max_memory=max_memory,
   (...)
    349         low_zero=(device_map == "balanced_low_0"),
    350     )
    351 if isinstance(device_map, str):
--> 352     device_map = infer_auto_device_map(
    353         model, max_memory=max_memory, no_split_module_classes=no_split_module_classes, dtype=dtype
...
--> 288     max_memory = {i: torch.cuda.mem_get_info(i)[0] for i in range(torch.cuda.device_count())}
    289 max_memory["cpu"] = psutil.virtual_memory().available
    290 return max_memory

AttributeError: module 'torch.cuda' has no attribute 'mem_get_info'
```

### Expected behavior

I followed the instruction advising me to upgrade torch to 1.9.0, which can enable `init_empty_weights`. However, torch 1.9.0 seems not to have `mem_get_info`. How to fix it? 

Thanks for any help!

**Comment by sgugger** [comment]
Looks like this function was only added in PyTorch 1.11, so you will need to upgrade to PyTorch 1.11 to be able to use this feature.

**Comment by BruceStayHungry** [comment]
Thanks! I further upgraded, and it works now

**Comment by BruceStayHungry** [comment]
It works on loading gpt-neox-20B. 
While unfortunately, when I use `load_checkpoint_and_dispatch` to load opt-30B, I set `device_map = "auto"`. It raised the following error:
```
ValueError: decoder.embed_tokens.weight doesn't have any device set.
```
I preliminarily debug it. It looks like that the auto-inferred `device_map` uses the key `decoder.embed_tokens`. But the `load_checkpoint_and_dispatch` use the key  `decoder.embed_tokens.weight`. Does the discrepancy cases the error? @sgugger 

The upgraded environments:
```
- `Accelerate` version: 0.14.0
- Platform: Linux-5.15.0-52-generic-x86_64-with-glibc2.17
- Python version: 3.8.15
- Numpy version: 1.23.5
- PyTorch version (GPU?): 1.14.0.dev20221122+cu117 (True)
- `Accelerate` default config:
        Not found
```

**Comment by BruceStayHungry** [comment]
My simple code:
```
model_name_or_path = "plms/opt-30b"

config = AutoConfig.from_pretrained(model_name_or_path)
config.use_cache = False

with init_empty_weights():
    model = AutoModelForCausalLM.from_config(config)
    tokenizer =  AutoTokenizer.from_pretrained(model_name_or_path, use_fast=False)


load_checkpoint_and_dispatch(
    model,
    model_name_or_path,
    device_map="auto",
    offload_folder=None,
    offload_state_dict=False,
    dtype="bfloat16"
)

print(model)

model = model.eval()
```

And the used [pytorch_model.bin.index.json](https://huggingface.co/facebook/opt-30b/blob/main/pytorch_model.bin.index.json)

**Comment by sgugger** [comment]
It won't work like this for OPT, you should use the `from_pretrained` method: the checkpoint only contains the base model while the model obtained with `AutoModelForCausalLM` will have more keys (like the decoder) which are tied, and also parameter names that do not match the checkpoint (since the model with the causal LM head has the base model as an attribute). All of that logic is in `from_pretrained`.

Your code would work if you use `AutoModel`, since in this case the model obtained would match the checkpoint.

**Comment by BruceStayHungry** [comment]
Thanks for your explanation. 

> the checkpoint only contains the base model while the model obtained with AutoModelForCausalLM will have more keys (like the decoder) which are tied

Would you mean that the released OPT-30b essentially doesn't contain the decoder part (convert hidden_state to tokens)?
Thus I can't use it like GPT2 to generate sentences. But the model card and its config.json show `OPTForCausalLM `. If so, 
how to load OPT to generate sentences?

**Comment by BruceStayHungry** [comment]
Well, I straightly use the `AutoModelForCausalLM.from_pretrained(model_name_or_path, device_map="auto")`  and it works for OPT.

**Comment by fssh72** [comment]
> It won't work like this for OPT, you should use the `from_pretrained` method: the checkpoint only contains the base model while the model obtained with `AutoModelForCausalLM` will have more keys (like the decoder) which are tied, and also parameter names that do not match the checkpoint (since the model with the causal LM head has the base model as an attribute). All of that logic is in `from_pretrained`.
> 
> Your code would work if you use `AutoModel`, since in this case the model obtained would match the checkpoint.

Hi @sgugger, for me the issue is that, when I use the default `AutoModelForCausalLM.from_pretrained` method with `device_map="auto"`, the model loads just fine on the GPUs, but when I start inference, I get a CUDA OOM error. By tweaking the `max_memory` map myself and using the `infer_auto_device_map` and `load_checkpoint_and_dispatch` I can get it to work. 

In this way I'm able to load, distribute and run `facebook/galactica-30b` on my hardware setup. However, for `facebook/opt-iml-max-30b` this doesn't work; I'm getting the same error:

> ValueError: decoder.embed_tokens.weight doesn't have any device set.

You indicated that the model should be loaded using `AutoModel`. My question is how to subsequently wrap that in a CausalLM model? Would using `OPTForCausalLM.from_config(config) instead work?

**Maybe on a higher level, my question is, how to use `infer_auto_device_map` and `load_checkpoint_and_dispatch` directly with `Facebook/opt-*` type of models, such that I can specify the `max_memory` map myself?**

Further, If the `device_map="auto"` would be a little bit more conservative, I would not have these issues at all. What can be improved there?

Thanks!

**Comment by sgugger** [comment]
> Maybe on a higher level, my question is, how to use infer_auto_device_map and load_checkpoint_and_dispatch directly with Facebook/opt-* type of models, such that I can specify the max_memory map myself?

You should use checkpoints that match the class of the model you want. If you don't want to use the `from_pretrained` method of Transformers that does all the magic adapation between the base model and the model with a head for you, I don't have anything to offer apart from looking at the source code and trying to replicate it yourself.

All in all this has nothing to do with Accelerate: if you try to load a state dict that doesn't match your model you get an error (as you would in PyTorch)

**Comment by fssh72** [comment]
Hi @sgugger, so I debug-stepped through `from_pretrained` to find what "magic" I needed to extract from it, and found that actually most of what is described on page ["Handling big models for inference"](https://huggingface.co/docs/accelerate/main/en/usage_guides/big_modeling#designing-a-device-map) has been used in `from_pretrained`, this was not clear to me from this page. Further, I found that one can provide `max_memory` to `from_pretrained` and that it is applied, even when `device_map="auto"`, which was also not clear from the documentation.

In conclusion, I could simply use `.from_pretrained`, providing a custom `max_memory` map, and all worked fine.

**Comment by usernaamee** [comment]
For anyone looking for an example, here is what works:

`import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

model = AutoModelForCausalLM.from_pretrained('./', device_map='auto', max_memory={0:'40GiB', 1:'40GiB', 'cpu':'256GiB'}) # Tweak the values as per your hardware
tokenizer = AutoTokenizer.from_pretrained('./', use_fast=False)
prompt = "What is the color of a carrot?\nA:"
input_ids = tokenizer(prompt, return_tensors="pt").input_ids.cuda()
generated_ids = model.generate(input_ids)
res = tokenizer.batch_decode(generated_ids, skip_special_tokens=True)
print(res)
`
